component/component.go

```diff
@@ -40,4 +56,17 @@ type Host interface {
 	// Context returns a context provided by the host to be used on the component
 	// operations.
 	Context() context.Context
+
+	// GetFactories of the specified kind. Returns a map of component types to factories.
+	// This allows components to create other components. For example:
+	//   func (r MyReceiver) Start(host component.Host) error {
+	//     factories := host.GetFactories(KindReceiver)
+	//     apacheFactory := factories["apache"].(receiver.Factory)
+	//     receiver, err := apacheFactory.CreateMetricsReceiver(...)
+	//     ...
+	//   }
+	// GetFactories can be called by the component anytime after Start() begins and
+	// until Shutdown() is called. Note that the component is responsible for destroying
+	// other components that it creates.
+	GetFactories(kind Kind) map[string]Factory
```

Comment:
map is mutable and some `smart caller` may change it. Maybe just `GetFactory(kind Kind, type string)`? or return just the []Factory slice - that also has the mutable problem up to a certain level?

Reply:
I like what you suggest. It is still one function but removes the mutation problem.

Reply:
I'll make this change.

Reply:
Done.